apps/google-analytics-4/frontend/src/services/api.ts

```diff
@@ -0,0 +1,103 @@
+import { z } from 'zod';
+import { config } from '../config';
+
+const ZCredentials = z.object({
+  status: z.string(),
+});
```

Comment:
Why are we validating the response in the client? We own both, the server and client, so we can rely on the shape to be fine

Reply:
I've seen many cases where even in the same team people make changes to a server while forgetting to make or deploy a change to the client, so in general it's a defensive approach. It's to avoid the `const someResult = getResult() as Result` class of runtime errors.

zod (new to me) gives both runtime schema and compile time type validation "for free." In general it would be even be more beneficial for user input of course.

My preference here would be to run with it just to learn from using it how valuable / not valuable it is. (I had in mind to refactor some handwritten type assertions I did for the credentials input.)

Reply:
I am totally fine with using a validation library for user input (e.g. credential validation).

I am worried about using a library to validate our server response. The front end might still work even though the validation fails. Example: we add a second property to the credentials status response. The whole application will stop working and the user cannot do anything about that.
Instead, I'd try to prevent this issue during compile time. Can we for example share TypeScript types between the lambda and the frontend? That way we can guarantee that the shape is correct.

Reply:
In general I think the point of validating an API response from a server is to fail loudly and explicitly vs. quietly and implicitly. Adding new attributes is not a requirement since these schemas are strict about what's included not about "extra." Removing attributes as with any migration could break things, but again I'd see this as a feature (you forgot to update your client that a given attribute was removed). It's a bit less clear the value in such a super trivial use case as this one but becomes more valuable as an API becomes more complex. The major downtime is the runtime cost of evaluating the schema on every API result, which is important to be sure.

Having said that, while I'm a strong believer in these benefits, it's not a dealbreaker requirement at this stage and I agree that we can take advantage of sharing the same codebase here to streamline this whole thing. The shared types is a nice idea I will look into now. 😄

Reply:
Hey let me know if you'd feel comfortable punting on this issue to a follow up PR.

I created a follow up ticket (INTEG-118) to address the issue of a shared schema (we're not using any kind of schema or types at all on the backend response). Sharing types would be super valuable but is non trivial to set up from a packaging perspective. 😢  But it's important so we definitely will circle back to find a good solution here.